### PR TITLE
Graceful shutdown of HTTP server and worker threads

### PR DESCRIPTION
**Pull request by frsyuki**
This change implements graceful shutdown to HTTP server and MultiThreadAgent threads.

* Undertow creates XnioWorker automatically but here doesn't use it. If we use it, `Undertow.stop` calls `xnioWorker.shutdownNow` immediately. However, it shouldn't be called until `GracefulShutdownHandler.shutodwn` is called to make sure that the HTTP handlers are not processing requests. Therefore, here creates XnioWorker and controls shutdown order manually.
* Upgraded Undertow to 1.4.x because `Undertow.Builder.setWorker` is not available in 1.3.
* A new shutdownHandler waits for completion of `ServerControl.destroy()`. It calls `@PreDestroy` methods. Thus shutting down digdag server will wait for completion of `@PreDestroy LocalAgentManager.shutdown` method.
* `LocalAgentManager.shutdown` calls `MultiThreadAgent.shutdown`. It waits for completion of all running tasks.

### CODE REVIEW

**Comment by frsyuki** [comment]
@danielnorberg PTAL
Multi-process agent is a next step that might not be implemented soon :-p

File: digdag-core/src/main/java/io/digdag/core/agent/MultiThreadAgent.java [hunk 1]
```diff
@@ -41,34 +44,56 @@ public MultiThreadAgent(AgentConfig config, AgentId agentId,
...
+        }
+        if (maximumCompletionWait.isPresent()) {
+            long seconds = maximumCompletionWait.get().getSeconds();
+            if (!executor.awaitTermination(seconds, TimeUnit.SECONDS)) {
+                logger.warn("Some tasks didn't finish within maximum wait time ({} seconds)", seconds);
+            }
+        }
+        else {
+            // no maximum wait time. waits for ever
+            while (!executor.awaitTermination(24, TimeUnit.HOURS))
```
Comment (danielnorberg) [inline digdag-core/src/main/java/io/digdag/core/agent/MultiThreadAgent.java]: Maybe log something to at least indicate why the agent is not terminating?

**Comment by danielnorberg** [comment]
Does this mean that the digdag rest api becomes unavailable until all currently running operators finish?

File: digdag-core/src/main/java/io/digdag/core/agent/MultiThreadAgent.java (hunk 1 above)
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/agent/MultiThreadAgent.java]: 👍  Added.

**Comment by frsyuki** [comment]
> Does this mean that the digdag rest api becomes unavailable until all currently running operators finish?

Yes. It's another reason why we should run agents on different machine.

**Comment by danielnorberg** [comment]
Ok, please add a test verifying that the rest api is available while the server waits for task completion.

**Comment by frsyuki** [comment]
@danielnorberg can I ask you to check new test?
The test runs only on unix systems because there're no good ways for Java to send signals to a PID on Windows.

**Comment by frsyuki** [comment]
Undertow 1.4.0.Final seems released but not on maven central yet somehow.
https://github.com/undertow-io/undertow/commits/1.4.x

Release note:
https://issues.jboss.org/secure/ReleaseNote.jspa?projectId=12312521&version=12330950

File: digdag-tests/src/test/java/acceptance/ServerGracefulShutdownIT.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                    "-c", config.toString(),
+                    "-e", server.endpoint(),
+                    "server_graceful_shutdown", "sleep",
+                    "--session", "now");
+            assertThat(startStatus.code(), is(0));
+            attemptId = getAttemptId(startStatus);
+        }
+
+        // Wait for the task to start
+        for (int i = 0; i < 30; i++) {
```
Comment (danielnorberg) [inline digdag-tests/src/test/java/acceptance/ServerGracefulShutdownIT.java]: Use `TestUtils.expect()` instead?

https://github.com/treasure-data/digdag/blob/master/digdag-tests/src/test/java/utils/TestUtils.java#L196

**Comment by danielnorberg** [comment]
👍

**Comment by frsyuki** [comment]
Now I could download undertow 1.4.0.Final, somehow... Updated.

File: digdag-tests/src/test/java/acceptance/ServerGracefulShutdownIT.java (hunk 2 above)
Comment (frsyuki) [inline digdag-tests/src/test/java/acceptance/ServerGracefulShutdownIT.java]: 👍  Updated.

**Comment by danielnorberg** [comment]
👍  LGTM

**Pull request merged**